File: omnibook-broker/src/main/java/com/sprint/omnibook/broker/domain/Room.java
package com.sprint.omnibook.broker.domain;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * 방 엔티티.
 */
@Entity
@Table(name = "room")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Room {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "property_id", nullable = false)
    private Property property;

    @Column(nullable = false)
    private String name;

    @Column(name = "room_type")
    private String roomType;

    private Integer capacity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RoomStatus status = RoomStatus.ACTIVE;

    @OneToMany(mappedBy = "room", cascade = CascadeType.ALL)
    private List<PlatformListing> platformListings = new ArrayList<>();

    @OneToMany(mappedBy = "room", cascade = CascadeType.ALL)
    private List<Reservation> reservations = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @Builder
    public Room(String name, String roomType, Integer capacity) {
        this.name = name;
        this.roomType = roomType;
        this.capacity = capacity;
    }

    @PrePersist
    protected void onCreate() {
        createdAt = Instant.now();
        updatedAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = Instant.now();
    }

    public void addPlatformListing(PlatformListing listing) {
        platformListings.add(listing);
        listing.assignRoom(this);
    }

    /**
     * 방을 숙소에 연결한다. (내부 사용 전용)
     */
    void assignProperty(Property property) {
        this.property = property;
    }
}

File: omnibook-broker/src/main/kotlin/com/sprint/omnibook/broker/domain/Room.kt
package com.sprint.omnibook.broker.domain

import jakarta.persistence.*
import java.time.Instant

/**
 * 방 엔티티.
 */
@Entity
@Table(name = "room")
class Room(
    @Column(nullable = false)
    var name: String,

    @Column(name = "room_type")
    var roomType: String? = null,

    var capacity: Int? = null
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "property_id", nullable = false)
    var property: Property? = null
        internal set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: RoomStatus = RoomStatus.ACTIVE

    @OneToMany(mappedBy = "room", cascade = [CascadeType.ALL])
    val platformListings: MutableList<PlatformListing> = mutableListOf()

    @OneToMany(mappedBy = "room", cascade = [CascadeType.ALL])
    val reservations: MutableList<Reservation> = mutableListOf()

    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: Instant? = null
        protected set

    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant? = null
        protected set

    @PrePersist
    protected fun onCreate() {
        createdAt = Instant.now()
        updatedAt = Instant.now()
    }

    @PreUpdate
    protected fun onUpdate() {
        updatedAt = Instant.now()
    }

    fun addPlatformListing(listing: PlatformListing) {
        platformListings.add(listing)
        listing.assignRoom(this)
    }

    /**
     * 방을 숙소에 연결한다. (내부 사용 전용)
     */
    internal fun assignProperty(property: Property) {
        this.property = property
    }
}
